### PR TITLE
(TK-162) sync ResolveSource and a test case for concatenation

### PR DESCRIPTION
**Pull request by KevinCorcoran**
This is done on top of #36, probably easiest to merge that one first.

### CODE REVIEW

**Comment by fpringvaldsen** [comment]
Looks good! Just had a few minor comments.

File: lib/hocon/impl/resolve_context.rb [hunk 1]
```diff
@@ -27,6 +29,32 @@ def self.new_cycle_markers
...
+    end
+    copy = self.class.new_cycle_markers
+    copy.merge(@cycle_markers)
+    copy.add(value)
+    self.class.new(@memos, @options, @restrict_to_child, @resolve_stack, copy)
+  end
+
+  def remove_cycle_marker(value)
+    if Hocon::Impl::ConfigImpl.trace_substitution_enabled
+      Hocon::Impl::ConfigImpl.trace("++ Cycle marker " + value + "@" + value.hash,
```
Comment (jpinsonault) [inline lib/hocon/impl/resolve_context.rb]: Probably not important, but I noticed that the java code says `-- Cycle marker`, with hyphens

**Comment by KevinCorcoran** [comment]
I'm going to close this PR and open a new one since the branch needs to be rebased onto master and there are comments on commits.

**Comment by KevinCorcoran** [comment]
@fpringvaldsen @jpinsonault - I just addressed your comments and opened #39 to replace this.

**Pull request closed without merging**